File: Episode03-Bullets_And_Collisions/Pygame/bullet.py
import pygame
import shared

class Bullet():
	def __init__(self, x:int, y:int, w:int, h:int, colour:tuple) -> None:
		self.rect = pygame.Rect(x, y, w, h)		# create a rectangle at 0,0 measuring w x h	
		self.colour = colour					# set colour
		self.rect.bottom = y
		self.rect.centerx = x
		self.speedy = 1000
		if shared.debug:
			self.speedy = 400
		self.active = True
		
	def get_rect(self):
		return self.rect

	def update(self, dt) -> None:
		self.rect.y -= round(self.speedy * dt)
		# delete if moves off top of screen
		if self.rect.bottom < 0:
			self.active = False
			
		return self.active
			
	def draw(self) -> None:
		pygame.draw.rect(shared.screen, self.colour, self.rect)

File: Episode05-Circle_Colliders_and_Animated_Sprites/Love2D/bullet.lua
local Class = require("lib.Class")
local Bullet = Class:derive("Bullet")
local Rectangle = require "lib.Rectangle"
local Circle = require "lib.Circle"
local Shared = require "shared"

function Bullet:new(Player, sprite)
	--[[ class constructor, takes x and y integer values and a sprite]]
	self.sprite = sprite
	self.rotation = 0 -- radians 360 deg = math.pi * 2 (approx 6.3 rad). 
	self.speedY = 1000
	if Shared.debug then
		self.speedY = 500
	end
	local posX = Player.rect.x + Player.rect.w / 2 - sprite:getWidth() / 2
	self.radius = math.min( sprite:getWidth() / 2, sprite:getHeight() / 2)
	-- rectangle collider
	self.rect = Rectangle(posX,
						  Player.rect.y - sprite:getHeight() + 10,
						  sprite:getWidth(),
						  sprite:getHeight())
	-- circle collider
	self.circle = Circle(self.rect.x + self.radius,
						 self.rect.y + self.radius,
						 self.radius)
	self.active = true
end

function Bullet:getCircle()
	return self.circle
end

function Bullet:getRect()
	return self.rect
end

function Bullet:update(dt)
	self.rect:updatePosition(0, self.speedY * dt * -1) -- negative as bullets go up
	self.circle:updatePosition(0, self.speedY * dt * -1)
	if self.rect.top <= 0 then
		self.active = false 
	end
	return self.active -- when false, bullet can be set to nil
end

function Bullet:draw()
	if Shared.debug then
		love.graphics.setColor(Shared.YELLOW)
		love.graphics.rectangle("line", self.rect.x, self.rect.y, self.rect.w, self.rect.h)
		love.graphics.setColor(Shared.RED)
		love.graphics.circle('line', self.circle.x, self.circle.y, self.circle.radius)
	else
		love.graphics.setColor(Shared.WHITE)
		love.graphics.draw(self.sprite, self.rect.x, self.rect.y)
	end
	
end

return Bullet